src/Microsoft.Restier.Core/ApiBase.cs

```diff
@@ -137,6 +131,31 @@ public void Dispose()
         }
 
         /// <summary>
+        /// Configure the entire service set for this API. Descendants can override this method to choose
+        /// from the default service sets provided by RESTier.
+        /// </summary>
+        /// <param name="apiType">
+        /// CLR type of this <see cref="ApiBase"/>.
+        /// </param>
+        /// <param name="services">
+        /// The <see cref="IServiceCollection"/> with which to create an <see cref="ApiConfiguration"/>.
+        /// </param>
+        /// <returns>The <see cref="IServiceCollection"/>.</returns>
+        [CLSCompliant(false)]
+        protected virtual IServiceCollection ConfigureApiServices(Type apiType, IServiceCollection services)
+        {
+            var customizer = ApiConfiguration.Customize(apiType);
+            services.Apply(customizer.InnerMost);
+
+            return this.ConfigureApi(services)
+                .Apply(customizer.PrivateApi)
+                .UseAttributes(apiType)
+                .UseConventions(apiType)
+                .Apply(customizer.Overrides)
+                .Apply(customizer.OuterMost);
+        }
+
```

Comment:
1. I will prefer we keep ApiBase. It is straight forward and easy to extend and understand.
2. Prefer to get rid of apiType here. Maybe make services as property of the class.
3. Prefer only one interface to override (maybe remove configureApi)
4. For end user who overrides the Api, we allows several ways to register,
a).  All user to add more service at the head, and if user uses cutoff here, it means he override the service.
base. ConfigureApiServices()
service.ChainPrevious()
service.cutoff

b).  All user to add more service at the tail, and if user can only use cutoff (or other like addInstance here), but can not use ChainPrevious here.
service.ChainPrevious()
service.cutoff()
base. ConfigureApiServices()

c). Allow user do at both side
service. cutoff();
base.Configure()
service.cutoff()
service.chainprevious()

d). If user wants more control, this is what he will do in override configureAPiService
var customizer = ApiConfiguration.Customize(apiType);
customizer.Inner.append()
customizer.Out.prepend()
base.configureApi()

comments?